hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -0,0 +1,399 @@
+                public void run() {
+                    try {
+                        Object value = resolve(state);
+                        if (value instanceof Throwable) {
+                            Throwable error = unwrap((Throwable) value);
+                            callback.onFailure(error);
+                        } else {
+                            callback.onResponse((V) value);
+                        }
+                    } catch (Throwable cause) {
+                        logger.severe("Failed asynchronous execution of execution callback: " + callback
+                                + "for call " + invocationToString(), cause);
+                    }
+                }
+
+            });
+        } catch (RejectedExecutionException e) {
+            logger.warning("Execution of callback: " + callback + " is rejected!", e);
+        }
+    }
+
+    // this method should not be needed; but there is a difference between client and server how it handles async throwables.
+    protected Throwable unwrap(Throwable throwable) {
+        if (throwable instanceof ExecutionException) {
+            return throwable.getCause();
+        }
+        return throwable;
+    }
+
+    protected abstract String invocationToString();
+
+    protected Object resolve(Object value) {
+        return value;
+    }
+
+    protected abstract V resolveAndThrow(Object state) throws ExecutionException, InterruptedException;
+
+    protected abstract TimeoutException newTimeoutException(long timeout, TimeUnit unit);
+
+    /**
+     * Registers a waiter (thread/ExecutionCallback) that gets notified when the future completes.
+     *
+     * @param waiter   the waiter
+     * @param executor the {@link Executor} to use in case of an {@link ExecutionCallback}.
+     * @return VOID if the registration was a success, anything else but void is the response.
+     */
+    private Object registerWaiter(Object waiter, Executor executor) {
+        WaitNode waitNode = null;
+        for (; ; ) {
+            final Object oldState = state;
+            if (isDone(oldState)) {
+                return oldState;
+            }
+
+            Object newState;
+            if (oldState == VOID && (executor == null || executor == defaultExecutor)) {
+                // Nothing is syncing on this future, so instead of creating a WaitNode, we just try to cas the waiter
+                newState = waiter;
+            } else {
+                // something already has been registered for syncing. So we need to create a WaitNode.
+                if (waitNode == null) {
+                    waitNode = new WaitNode(waiter, executor);
+                }
+                waitNode.next = oldState;
+                newState = waitNode;
+            }
+
+            if (compareAndSetState(oldState, newState)) {
+                // we have successfully registered
+                return VOID;
+            }
+        }
+    }
+
+    void unregisterWaiter(Thread waiter) {
```

Comment:
is this method needed at all? it's quite complex and it mingles with the (concurrent) waiting list -> it's error-pron

However what are consequences if you do not un-register a waiter at all? a thread with receive a permit at some point in time. but that is absolutely OK as spurious permits/wake-up are normal.